landscape/client/manager/snapmanager.py

```diff
@@ -0,0 +1,261 @@
+        where each task must be performed independently per-snap.
+
+        This is required when we want to provide metadata for refreshes
+        or installs and also specify the channel, revision, or other
+        arguments per-snap.
+        """
+        logging.debug(f"Handling message {message} as multiple snap tasks")
+        message_type = message["type"]
+        snaps = message["snaps"]
+        opid = message["operation-id"]
+        errors = {}
+        queue = deque()
+
+        logging.info(f"Performing {message_type} action for snaps {snaps}")
+
+        # Naively doing this synchronously because each is an HTTP call to the
+        # snap REST API that returns basically immediately. We poll for their
+        # completion statuses once they've all been kicked off.
+        for snap in snaps:
+            name = snap["name"]
+            snap_args = snap.get("args", {})
+
+            try:
+                response = self._start_snap_task(
+                    message_type,
+                    name,
+                    **snap_args,
+                )
+                queue.append((response["change"], name))
+            except SnapdHttpException as e:
+                result = e.json["result"]
+                logging.error(
+                    f"Error in {message_type} for '{name}': {message}",
+                )
+                errors[name] = result
+
+        deferred = self._check_statuses(queue)
+        deferred.addCallback(self._respond, opid, errors)
+
+        return deferred
+
+    def _check_statuses(self, change_queue):
+        """
+        Repeatedly polls for the status of each change in `change_queue`
+        until all are no longer in-progress.
+        """
+        completed_changes = []
+        interval = getattr(self.registry.config, "snapd_poll_interval", 15)
+
+        def get_status():
+            """
+            Looping function that polls snapd for the status of
+            changes, moving them from the queue when they are done.
+            """
+            if not change_queue:
+                loop.stop()
+                return
+
+            logging.info("Polling snapd for status of pending snap changes")
+
+            try:
+                result = self._snap_http.check_changes().get("result", [])
+            except SnapdHttpException as e:
+                logging.error(f"Error checking status of snap changes: {e}")
+                completed_changes.extend(
+                    [(name, str(e)) for _, name in change_queue],
+                )
+                loop.stop()
+                return
+
+            for _ in range(len(change_queue)):
+                cid, name = change_queue.popleft()
+
+                # It's possible (though unlikely) that a change is not in the
+                # list - snapd could have dropped it for some reason. We need
+                # to know if that happens, hence the extra `found` check.
+                found = False
```

Comment:
Use the found flag and breaks in the double loop is a bit confusing. I think it would be best to define result_dict on line 155 which would be the same as result but keyed by id. Then eliminate the second loop and check in there if it exists or not

Reply:
I like that. I'll retool it to avoid the extra loop and nesting.